zetaclient/chains/zrepo/zrepo.go

```diff
@@ -37,16 +88,257 @@ func (repo *ZetaRepo) GetOutboundTrackers(ctx context.Context) ([]types.Outbound
+// Voting & Posting Trackers
+// ------------------------------------------------------------------------------------------------
+
+// PostOutboundTracker posts an outbound tracker.
+// It returns the hash of the associated ZetaChain transaction.
+func (repo *ZetaRepo) PostOutboundTracker(ctx context.Context, logger zerolog.Logger,
+	nonce uint64,
+	txHash string,
+) (string, error) {
+	// Does not post outbound trackers in dry mode.
+	if repo.clientMode == mode.DryMode {
+		logger.Info().Stringer(logs.FieldMode, mode.DryMode).Msg("skipping outbound tracker")
+		return "", nil
+	}
+
+	zhash, err := repo.client.PostOutboundTracker(ctx, repo.connectedChain.ChainId, nonce, txHash)
+	if err != nil {
+		err = newClientError(ErrClientPostOutboundTracker, err)
+		logger.Error().Err(err).Send()
+		return "", err
+	}
+
+	if zhash == "" {
+		logger.Info().Msg("outbound tracker already exists")
+	} else {
+		logger.Info().Str(logs.FieldZetaTx, zhash).Msg("added outbound tracker")
+	}
+
+	return zhash, nil
+}
+
+// VoteGasPrice votes on gas prices.
+// It returns the hash of the vote transaction.
+func (repo *ZetaRepo) VoteGasPrice(ctx context.Context, logger zerolog.Logger,
+	gasPrice uint64,
+	priorityFee uint64,
+	block uint64,
+) (string, error) {
+	// Does not vote in dry mode.
+	if repo.clientMode == mode.DryMode {
+		logger.Info().Stringer(logs.FieldMode, mode.DryMode).Msg("skipping gas price vote")
+		return "", nil
+	}
+
+	zhash, err := repo.client.PostVoteGasPrice(ctx, repo.connectedChain, gasPrice, priorityFee, block)
+	if err != nil {
+		err = newClientError(ErrClientVoteGasPrice, err)
+		logger.Error().Err(err).Send()
+		return "", err
+	}
+	return zhash, nil
+}
+
+// VoteInbound votes on an inbound.
+// It skips invalid messages and tries to vote on the ballot even if the inbound already has a CCTX.
+// It returns the index of the ballot.
+func (repo *ZetaRepo) VoteInbound(ctx context.Context, logger zerolog.Logger,
+	msg *crosschain.MsgVoteInbound,
+	retryGasLimit uint64,
+) (string, error) {
+	logger = logger.With().
+		Str(logs.FieldTx, msg.InboundHash).
+		Stringer(logs.FieldCoinType, msg.CoinType).
+		Stringer("confirmation_mode", msg.ConfirmationMode).
+		Logger()
+
+	{
+		cctxIndex := msg.Digest()
+
+		// A CCTX is created after an inbound ballot is finalized.
+		// - If the CCTX already exists, we try voting if the finalized ballot is still present.
+		// - If the CCTX exists but the ballot does not exist, we do not vote.
+		cctxExists, err := repo.cctxExists(ctx, cctxIndex)
+		if err != nil {
+			return "", err
+		}
+		if cctxExists {
+			ballotExists, err := repo.ballotExists(ctx, cctxIndex)
+			if err != nil {
+				return "", err
+			}
+
+			if !ballotExists {
+				logger.Info().Msg("not voting on inbound; CCTX exists but the ballot does not")
+				return cctxIndex, nil
+			}
+		}
+	}
+
+	// Validates the message to avoid unnecessary retries.
+	err := msg.ValidateBasic()
+	if err != nil {
```

Comment:
NIT: we can use one-liner if

```suggestion
	if err := msg.ValidateBasic(); err != nil {
```